docs/userguide/app_caching.rst

```diff
@@ -3,56 +3,60 @@
 App caching
 ----------
 
-When developing a workflow, developers often run the same workflow
-with incremental changes over and over. Often large fragments of
-a workflow will not have changed, yet apps will be re-executed, wasting
-valuable developer time and computation resources. App caching
-solves this problem by storing results from apps that have completed
-so that they can be re-used. App caching can be enabled by setting the ``cache``
-argument in the :func:`~parsl.app.app.python_app` or :func:`~parsl.app.app.bash_app` decorator to ``True`` (by default it is ``False``). App caching
-can be globally disabled by setting ``app_cache=False``
-in the :class:`~parsl.config.Config`.
+There are many situations in which a program may be re-executed
+over time. Often large fragments of the program will not have changed, 
+and therefore, re-execution of apps will waste valuable time and 
+computation resources. Parsl's app caching solves this problem by 
+storing results from apps that have successfully completed
+so that they can be re-used. 
+
+App caching is enabled by setting the ``cache``
+argument in the :func:`~parsl.app.app.python_app` or :func:`~parsl.app.app.bash_app` 
+decorator to ``True`` (by default it is ``False``). 
 
 .. code-block:: python
 
    @bash_app(cache=True)
    def hello (msg, stdout=None):
        return 'echo {}'.format(msg)
+			
+App caching can be globally disabled by setting ``app_cache=False``
+in the :class:`~parsl.config.Config`.
 
-
-App caching can be particularly useful when developing interactive workflows such as when
+App caching can be particularly useful when developing interactive programs such as when
 using a Jupyter notebook. In this case, cells containing apps are often re-executed
 during development. Using app caching will ensure that only modified apps are re-executed.
 
 
-Inputs
-^^^^^^
+App equivalence 
+^^^^^^^^^^^^^^^
 
 Two app invocations are treated as the same by the caching mechanism if their
-inputs are the same. This sameness is determined by hashing the inputs, and
-comparing hashes.
+input arguments are the same. This equivalence is determined by hashing the app's input
+arguments, and comparing hashes. 
```

Comment:
Is there also a check on the app's content?

Reply:
That's something of a rathole and there is at least this open issue about it, #1591
 
The answer, at least for master, is "sort of". (The direction I will be pushing after this release is "no").

The python function definition is hashed. If you change the function definition then you'll change the hash.

If you change some probably irrelevant parts of the function definition you will change the hash (such as line number) - for example, adding a one line #comment at the top of your source file invalidates all checkpoints, because it increments the line number for each function by 1.

If you change a helper function that you're calling from the function definition, you won't change the hash.

I intend this to change by 1.1 because this is confusing.

A short honest summary is "changing the source code of your workflow/program will sometimes invalidate checkpoints and sometimes not".